File: filter.go
package main

import f "fmt"
import j "encoding/json"
import s "strings"

type Branch struct {
	Name string
	Branches []Branch
}

//callback which decides is branch applied with filter or not..
func filterCallback(branch Branch, search []string) bool {
	for i := 0; i < len(search); i++ {
		if s.Contains(branch.Name, search[i]) {	
			return true
		}
	}

	return false
}

//filter tree recursively based on provided search filter
func filter(parentBranch Branch, search []string) (bool, Branch) {
	if len(parentBranch.Branches) > 0 {
		var branches []Branch

		for i := 0; i < len(parentBranch.Branches); i++ {
			applyToFilter, filteredBranch := filter(parentBranch.Branches[i], search);

			if applyToFilter == true {
				branches = append(branches, filteredBranch)
			}
		}

		if len(branches) > 0 {
			parentBranch.Branches = branches
			return true, parentBranch
		}

		return false, parentBranch
	}

	return filterCallback(parentBranch, search), parentBranch
}

func main() {

	//init tree and create json in string
	var rootBranch Branch
	branchesJson := `{"name":"Root","branches":[{"name":"Root Branch 1","branches":[{"name":"Branch 2","branches":[]},{"name":"Branch 3","branches":[{"name":"Branch 4","branches":[]},{"name":"Branch 5","branches":[]}]}]},{"name":"Root Branch 2","branches":[{"name":"Branch 1","branches":[{"name":"Branch 2","branches":[{"name":"Special Branch","branches":[]}]},{"name":"One more specific branch","branches":[]}]}]}]}`

	//parse json to create tree
	j.Unmarshal([]byte(branchesJson), &rootBranch)

	var branches []Branch
	search := []string{"One"}

	for i := 0; i < len(rootBranch.Branches); i++ {
		applyToFilter,filteredBranch := filter(rootBranch.Branches[i], search);

		if applyToFilter == true {
			branches = append(branches, filteredBranch)
		}
	}

	rootBranch.Branches = branches

	f.Println(rootBranch)
}

File: filter.js
let tree = {
	name: "Root",
	branches: [
	{
			name: "Root Branch 1",
			branches: [
				{
					name: "Branch 2",
					branches: [],
				},
				{
					name: "Branch 3",
					branches: [
						{
							name: "Branch 4",
							branches: [],
						},
						{
							name: "Branch 5",
							branches: []
						}
					]
				},
			]
		},
		{
			name: "Root Branch 2",
			branches: [
				{
					name: "Branch 1",
					branches: [
						{
							name: "Branch 2",
							branches: [
								{
									name: "Special Branch",
									branches: [],
								}
							]
						},
						{
							name: "One more specific branch",
							branches: [],
						}
					]
				},
			]
		}
	]
}

//filter callback which decides if branch is apply with search filter
function filterCallback(branch, search) {
	if (Array.isArray(search)) {
		let i = 0

		while(search[i]) {
			if (branch.name.search(search[i] + "") != -1)
				return true;

			i++
		}

		return false
	}

	return branch.name.search(search) != -1
}

//filter recursibely to find all branches which apply with search filter
function filter(parentBranch, search) {
	if (parentBranch.branches.length > 0) {
		parentBranch.branches = parentBranch.branches.filter(childBranch => filter(childBranch, search))

		if (filterCallback(parentBranch, search) || parentBranch.branches.length > 0)
			return parentBranch;

		return false;
	}

	return filterCallback(parentBranch, search)
}

//set search filter
const search = ['One']

//check recursively all branches and filter them
tree.branches = tree.branches.filter(branch => filter(branch, search));

console.log(JSON.stringify(tree));

